// skydapp-common/lib/SkyUtil.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class SkyUtil {
    static pull(array, ...removeList) {
        for (const el of removeList) {
            const index = array.indexOf(el);
            if (index !== -1) {
                array.splice(index, 1);
            }
        }
    }
    static insert(array, index, item) {
        array.splice(index, 0, item);
    }
    static random(min, max) {
        return Math.floor(Math.random() * (max - min + 1) + min);
    }
    static repeat(times, func) {
        for (let i = 0; i < times; i += 1) {
            if (func(i) === false) {
                break;
            }
        }
    }
    static repeatResult(times, func) {
        const results = [];
        for (let i = 0; i < times; i += 1) {
            results.push(func(i));
        }
        return results;
    }
    static async repeatResultAsync(times, func) {
        const results = [];
        for (let i = 0; i < times; i += 1) {
            results.push(await func(i));
        }
        return results;
    }
    static toTitleCase(str) {
        return str.replace(/(^\w|\s\w)(\S*)/g, (_, m1, m2) => m1.toUpperCase() + m2.toLowerCase());
    }
    static shortenAddress(address) {
        return `${address.substring(0, 6)}...${address.substring(38)}`;
    }
    static numberWithCommas(x, fixed = 3) {
        const parts = String(+(+x).toFixed(fixed)).split(".");
        parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, ",");
        return parts.join(".");
    }
}
exports.default = SkyUtil;
